musig/README.md

```diff
@@ -123,24 +167,52 @@ There are several paths to signing:
 
     For more information on each of these states and steps, see the [protocol for party state transitions](#protocol-for-party-state-transitions).
 
+3. Make a Schnorr signature with multiple public keys and multiple messages, in a way that is safe from Russell's attack.
+    - Create a `Multimessage` context by calling `Multimessage::new(...)`. 
+      See the [multimessage](#multimessage) section for more details.
+
+    For each party that is taking part in the signing:
+    - Call `Party::new(transcript, privkey, multimessage)`.
+    - All following steps are the same as in protocol #2, making a Schnorr signature with one aggregated key.
+
 ### Verifying
 
-Signature verification happens in the `Signature::verify(...)` function.
+There are several paths to verifying: 
+1. Normal Schnorr signature verification (covers cases #1 and #2 in [signing section](#signing)).
+    Function: `Signature::verify(...)`
 
-Input: 
-- `&self`
-- transcript: `&mut Transcript` - a transcript to which the signed message has already been committed.
-- P: `VerificationKey`
+    Input: 
+    - `&self`
+    - transcript: `&mut Transcript` - a transcript to which the signed message has already been committed.
+    - P: `VerificationKey`
 
-Operation:
-- Make `c = H(X, R, m)`. Since the transcript already has the message `m` committed to it, 
-the function only needs to commit `X` with label "X" and `R` with label "R", 
-and then get the challenge scalar `c` with label "c".
-- Decompress verification key `P`. If this fails, return `Err(VMError::InvalidPoint)`.
-- Check if `s * G == R + c * P`. `G` is the [base point](#base-point).
+    Operation:
+    - Make `c = H(X, R, m)`. Since the transcript already has the message `m` committed to it, 
+    the function only needs to commit `X` with label "X" and `R` with label "R", 
+    and then get the challenge scalar `c` with label "c".
+    - Decompress verification key `P`. If this fails, return `Err(VMError::InvalidPoint)`.
+    - Check if `s * G == R + c * P`. `G` is the [base point](#base-point).
 
-Output:
-- `Ok(())` if verification succeeds, or `Err(VMError)` if the verification or point decompression fail.
+    Output:
+    - `Ok(())` if verification succeeds, or `Err(VMError)` if the verification or point decompression fail.
+
+2. Multi-message Schnorr signature verification (covers case #3 in [signing section](#signing)).
+    Function: `Signature::verify_multimessage(...)`
+
+    Input: 
+    - `&self`
+    - transcript: `&mut Transcript` - a transcript to which the signed message has already been committed.
+    - multimessage: `Multimessage` 
+
+    Operation:
+    - Iterate through each (pubkey, message) pair in `multimessage`. For each, create `c_i` by calling
+      `multimessage.challenge(...)` and passing in the verification key, transcript, and `self.R`. 
```

Comment:
per notes above, how about this:

1. use `multimessage.commit(&mut t)` to commit the key
2. commit the R directly w/o using multimessage: `t.commit(R)`
3. use `multimessage.challenge(pk, &mut t)` to squeeze the per-point challenge.
4. Sum up `sum_i(X_i * c_i)` into `cX` and then check `s*G == cX + R`.

Reply:
done, thanks for making that clearer!